src/main/java/redis/clients/jedis/JedisClusterInfoCache.java

```diff
@@ -175,68 +177,72 @@ public void discoverClusterNodesAndSlots(Jedis jedis) {
 
   public void renewClusterSlots(Jedis jedis) {
     // If rediscovering is already in process - no need to start one more same rediscovering, just return
-    if (!rediscovering) {
+    if (rediscoverLock.tryLock()) {
       try {
-        w.lock();
-        if (!rediscovering) {
-          rediscovering = true;
-
+        if (jedis != null) {
           try {
-            if (jedis != null) {
-              try {
-                discoverClusterSlots(jedis);
-                return;
-              } catch (JedisException e) {
-                // try nodes from all pools
-              }
-            }
+            discoverClusterSlots(jedis);
+            return;
+          } catch (JedisException e) {
+            // try nodes from all pools
+          }
+        }
 
-            for (JedisPool jp : getShuffledNodesPool()) {
-              Jedis j = null;
-              try {
-                j = jp.getResource();
-                discoverClusterSlots(j);
-                return;
-              } catch (JedisConnectionException e) {
-                // try next nodes
-              } finally {
-                if (j != null) {
-                  j.close();
-                }
-              }
-            }
+        List<JedisPool> shuffledNodesPool = getShuffledNodesPool();
+        // Lower priority of last node which has issue to retrieve cluster slots
+        if (jedis != null) {
+          Pool<Jedis> dataSource = jedis.getDataSource();
+          if (shuffledNodesPool.remove(dataSource)) {
+            shuffledNodesPool.add((JedisPool) dataSource);
+          }
+        }
```

Comment:
@Shawyeok It seems unnecessary, because if `jedis != null`, it means that `JedisRedirectionException` was caught before (not `JedisConnectionException)`, then we use it call `discoverClusterSlots` will also success.

Reply:
Right, if cut this, then `Jedis#getDataSource()` is no need too.